ENH: Multi-level join on multi-indexes

harisbal opened the pull request:
closes #6360

Allow for merging on multiple levels of multi-indexes

gfyoung commented on pandas/tests/reshape/test_merge.py:
```diff
@@ -1136,14 +1136,14 @@ def test_join_multi_levels(self):
             household.join(portfolio, how='inner')
-        pytest.raises(ValueError, f)
+        self.assertRaises(ValueError, f)
 
```
Do not change this.  We're using `pytest` and not `nosetest`.

gfyoung commented on pandas/tests/reshape/test_merge.py:
```diff
@@ -1136,14 +1136,14 @@ def test_join_multi_levels(self):
             portfolio2.join(portfolio, how='inner')
-        pytest.raises(ValueError, f)
+        self.assertRaises(ValueError, f)
 
```
Same as <a href="https://github.com/pandas-dev/pandas/pull/16162/files#r114061443">here</a>.

gfyoung commented on pandas/tests/reshape/test_merge.py:
```diff
@@ -1208,12 +1204,179 @@ def f():
+                
+        def f():
+            matrix.join(distances2, how='left')
+        self.assertRaises(TypeError, f)
```
Same as <a href="https://github.com/pandas-dev/pandas/pull/16162/files#r114061443">here</a>.

gfyoung commented on pandas/tests/reshape/test_merge.py:
```diff
@@ -1208,12 +1204,179 @@ def f():
+        
+        def f():
+            matrix.join(distances2, how='left')
+        self.assertRaises(ValueError, f)
```
Same as <a href="https://github.com/pandas-dev/pandas/pull/16162/files#r114061443">here</a>.

harisbal commented:
Can anyone give some insight on the failure of Travis CI tests?

TomAugspurger commented:
Some linting (style) issues, starting at https://travis-ci.org/pandas-dev/pandas/jobs/232494137#L1403

You can `pip install flake8` and run it on the files your changing.

jreback commented on pandas/core/indexes/base.py:
```diff
@@ -3028,27 +3028,84 @@ def join(self, other, how='left', level=None, return_indexers=False,
+
+            return join_index, lidx, ridx
+
+        else:
```
you don't need the else here

jreback commented on pandas/tests/reshape/test_merge.py:
```diff
@@ -1195,7 +1191,7 @@ def f():
             DataFrame(dict(
                 household_id=[1, 2, 2, 2, 2, 3, 3, 3, 3, 3, 3, 4],
-                asset_id=["nl0000301109", "nl0000289783", "gb00b03mlx29",
+                asset_id=["nl0000301109", "nl0000301109", "gb00b03mlx29",
```
why is this changed?

harisbal replied:
I believe it was a mistake in the first instance. The household dataframe in this method suggests this correction

jreback replied:
why does the existing code not fail then? before this change

harisbal replied:
It was raising a NotImplementedError

jreback commented on pandas/tests/reshape/test_merge.py:
```diff
@@ -1208,11 +1204,179 @@ def f():
+
+        assert_frame_equal(result, expected)
+
+    def test_join_multi_levels3(self):
```
where are the error cases, e.g. mistmatch on multi-levels? misnamed levels.

harisbal replied:
https://github.com/harisbal/pandas/blob/4ed307074a4c376db209b44aae5b7eef00276d26/pandas/tests/reshape/test_merge.py#L1339-L1352

jreback commented on pandas/core/indexes/base.py:
```diff
@@ -3028,27 +3028,84 @@ def join(self, other, how='left', level=None, return_indexers=False,
+            other_tmp = other.droplevel(rdrop_levels)
+
+            if not (other_tmp.is_unique and self_tmp.is_unique):
+                raise TypeError(" The index resulting from the overlapping "
```
need testing for this

harisbal replied:
https://github.com/harisbal/pandas/blob/4ed307074a4c376db209b44aae5b7eef00276d26/pandas/tests/reshape/test_merge.py#L1323-L1337

jreback commented on pandas/core/indexes/base.py:
```diff
@@ -3028,27 +3028,84 @@ def join(self, other, how='left', level=None, return_indexers=False,
-        self_is_mi = isinstance(self, MultiIndex)
-        other_is_mi = isinstance(other, MultiIndex)
+
+        def _complete_join():
```
this should be a separate method, too confusing here

harisbal reviewed (commented):
Changes implemented

jreback commented on pandas/core/indexes/base.py:
```diff
@@ -3028,27 +3028,61 @@ def join(self, other, how='left', level=None, return_indexers=False,
         # figure out join names
         self_names = [n for n in self.names if n is not None]
         other_names = [n for n in other.names if n is not None]
         overlap = list(set(self_names) & set(other_names))
```
this should remain a set

jreback commented on pandas/core/indexes/base.py:
```diff
@@ -3028,27 +3028,61 @@ def join(self, other, how='left', level=None, return_indexers=False,
-        # make the indices into mi's that match
-        if not (self_is_mi and other_is_mi):
+            if not (self_tmp.is_unique and other_tmp.is_unique):
+                raise TypeError(" The index resulting from the overlapping "
```
use the original error message and it should be a ValueError

jreback commented on pandas/core/indexes/base.py:
```diff
@@ -3028,27 +3028,61 @@ def join(self, other, how='left', level=None, return_indexers=False,
+                join_index = self
+            elif how == 'right':
+                join_index = other
+            else:
```
else is unecessary

jreback commented on pandas/core/indexes/base.py:
```diff
@@ -3028,27 +3028,61 @@ def join(self, other, how='left', level=None, return_indexers=False,
+
+            return join_index, lidx, ridx
+        else:
+            jl = overlap[0]
```
don't need the else here as you are returning

jreback commented on pandas/core/indexes/base.py:
```diff
@@ -3028,27 +3028,61 @@ def join(self, other, how='left', level=None, return_indexers=False,
+        ldrop_lvls = [l for l in self_names if l not in overlap]
+        rdrop_lvls = [l for l in other_names if l not in overlap]
+
+        self_is_mi = isinstance(self, MultiIndex)
```
instead of doing this inline it should be a separate method, its too long

jreback commented on pandas/core/indexes/base.py:
```diff
@@ -3065,9 +3099,30 @@ def _join_multi(self, other, how, return_indexers=True):
-        raise NotImplementedError("merging with both multi-indexes is not "
-                                  "implemented")
+    def _complete_multi_join(self, other, join_idx, lidx, ridx, dropped_lvls):
+        new_lvls = join_idx.levels
```
doc-string on what this is

jreback commented on pandas/core/indexes/base.py:
```diff
@@ -3065,9 +3099,30 @@ def _join_multi(self, other, how, return_indexers=True):
+        for n in dropped_lvls:
+            if n in self.names:
+                idx = lidx
+                lvls = self.levels[self.names.index(n)].values
```
don't abbreviate on the variable names, use new_levels and new_labels etc

jreback commented on pandas/tests/reshape/test_merge.py:
```diff
@@ -1208,11 +1204,179 @@ def f():
+        result = matrix.join(distances, how='outer')
+        assert_frame_equal(result, expected)
+
+        # Non-unique resulting index
```
separate test for non-uniques

jreback commented on pandas/tests/reshape/test_merge.py:
```diff
@@ -1208,11 +1204,179 @@ def f():
+                     Destination=[1, 2, 1, 3, 1],
+                     Period=['AM', 'PM', 'IP', 'AM', 'OP'],
+                     TripPurp=['hbw', 'nhb', 'hbo', 'nhb', 'hbw'],
+                     Trips=[1987, 3647, 2470, 4296, 4444],
```
in fact why don't you move all of these added tests to a new TestMultiMulti class for testing (and other relevant tests if any).

jreback commented:
pls rebase and update

pep8speaks commented:
Hello @harisbal! Thanks for updating the PR.

 - In the file [`pandas/core/indexes/base.py`](https://github.com/pandas-dev/pandas/blob/763c892179721cf8b78cb7024092a9e17d2866fb/pandas/core/indexes/base.py), following are the PEP8 issues :

> [Line 3208:13](https://github.com/pandas-dev/pandas/blob/763c892179721cf8b78cb7024092a9e17d2866fb/pandas/core/indexes/base.py#L3208): [E741](https://duckduckgo.com/?q=pep8%20E741) ambiguous variable name 'l'

harisbal reviewed (commented):
Rebased

TomAugspurger reviewed (approved):
cc @jreback

jreback commented on pandas/core/indexes/base.py:
```diff
@@ -3111,46 +3111,125 @@ def join(self, other, how='left', level=None, return_indexers=False,
+            other_tmp = other.droplevel(rdrop_levels)
+
+            if not (self_tmp.is_unique and other_tmp.is_unique):
+                raise ValueError("Join on level between two MultiIndex objects"
```
test for this?

jreback commented on pandas/core/indexes/base.py:
```diff
@@ -3111,46 +3111,125 @@ def join(self, other, how='left', level=None, return_indexers=False,
+                elif how == 'right':
+                    join_index = other
+
+                levels = join_index.levels
```
move these 3 assignments after the .join, then you can make the 'how' an if/elif (for each of outer/left/right)

jreback commented on pandas/core/indexes/base.py:
```diff
@@ -3111,46 +3111,125 @@ def join(self, other, how='left', level=None, return_indexers=False,
-                if isinstance(result, tuple):
-                    return result[0], result[2], result[1]
-            return result
+        flip_order = False
```
huh?

if you need to do something based on the how, then do it above

harisbal replied:
The part from `jl = list(overlap)[0]` and onwards refers to cases where not both indexes are multiindexes.
This is legacy code that I didn't modify. Shall I try to rewrite?

jreback commented on pandas/core/indexes/base.py:
```diff
@@ -3111,46 +3111,125 @@ def join(self, other, how='left', level=None, return_indexers=False,
+        result = self._join_level(other, level, how=how,
+                                  return_indexers=return_indexers)
+
+        if flip_order:
```
this is non-intuitive here, have no idea what is going one

harisbal replied:
Same as above. Legacy code that I didn't modify

jreback commented on pandas/core/indexes/base.py:
```diff
@@ -3111,46 +3111,125 @@ def join(self, other, how='left', level=None, return_indexers=False,
+                return result[0], result[2], result[1]
+        return result
+
+    def _complete_outer_join(self, other, join_idx, lidx, ridx,
```
move to this pandas/core/reshape/merge.py (below _get_join_indexers) and just import where you need

jreback commented on pandas/tests/reshape/test_merge.py:
```diff
@@ -1344,11 +1341,191 @@ def f():
+        # Multi-index join tests
+        # Self join
+        matrix = (
+            pd.DataFrame(
```
pls parametrize or use fixtures for this.

jreback commented on pandas/tests/reshape/test_merge.py:
```diff
@@ -1344,11 +1341,191 @@ def f():
+
+        assert_frame_equal(result, expected)
+
+    def test_join_multi_levels3(self):
```
rename to something more useful

jreback reviewed (changes requested):
a number of comments

jreback commented:
also the docs in merging.rst need updating. pls add a small sub-section in 0.21.0 as well showcasing the new functionaility.

harisbal commented:
Thank you @jreback, I'll implement the changes

jreback commented:
@harisbal great. sorry it took so long on this.

harisbal commented:
It was actually my fault. Thanks again for all the comments :)

jreback commented on pandas/core/reshape/merge.py:
```diff
@@ -1066,6 +1066,54 @@ def _get_join_indexers(left_keys, right_keys, sort=False, how='inner',
     return join_func(lkey, rkey, count, **kwargs)
 
 
+def _complete_outer_join(self, other, join_idx, lidx, ridx, dropped_levels):
```
call these left/right (not self, other), add to doc-string

jreback commented on pandas/core/reshape/merge.py:
```diff
@@ -1066,6 +1066,54 @@ def _get_join_indexers(left_keys, right_keys, sort=False, how='inner',
+    Complete the index in case of outer join
+
+    Parameters
+    ----------
```
document like
```
join_index : Index
    the index of the join between common leves
lidx : intp array
    left indexer
....
```

etc

jreback commented on pandas/core/indexes/base.py:
```diff
@@ -3111,27 +3111,64 @@ def join(self, other, how='left', level=None, return_indexers=False,
+            dropped_lvls = ldrop_levels + rdrop_levels
+
+            # tmp_index is equivalent of index when how='inner'
+            tmp_index, lidx, ridx = self_tmp.join(other_tmp, how=how,
```
don't use ``tmp_index`` call this something else (and self_tmp)

harisbal replied:
How about:
self_tmp --> left
other_tmp --> right
tmp_index --> inner_join_index

jreback commented on pandas/core/indexes/base.py:
```diff
@@ -3111,27 +3111,64 @@ def join(self, other, how='left', level=None, return_indexers=False,
+            # Append to the returned Index the non-overlapping levels
+            dropped_lvls = ldrop_levels + rdrop_levels
+
+            # tmp_index is equivalent of index when how='inner'
```
explicity enumerate all of the how cases

```
if how == 'innter':
     pass
elif how == 'outer':
    # outer_join
elif how == 'left'
.....
```

jreback commented on doc/source/merging.rst:
```diff
@@ -1120,6 +1120,13 @@ This is not Implemented via ``join`` at-the-moment, however it can be done using
           labels=['left', 'right'], vertical=False);
    plt.close('all');
 
+For previous versions can be done using the following.
```
rather I would say this is equivalent to

jreback commented on doc/source/merging.rst:
```diff
@@ -1098,7 +1098,8 @@ This is equivalent but less verbose and more memory efficient / faster than this
 
-This is not Implemented via ``join`` at-the-moment, however it can be done using the following.
+.. versionadded:: 0.21.0
+As of version 0.21.0 joining on two multi-indexes is possible:
```
can you add a similar example to a sub-section in whatsnew (and put a ref to here). you can use this example and show how we *could* do it previously (e.g. the below section), and how it will just work now.

harisbal replied:
Shall I wait to include it in 0.21.1?

jreback replied:
going to be for next version (so you can wait to move it is ok), after we tag can merge shortly after.

jreback commented on pandas/core/indexes/base.py:
```diff
@@ -3111,27 +3111,53 @@ def join(self, other, how='left', level=None, return_indexers=False,
+
+            return multilevel_join_index, lidx, ridx
+
         if not (self_is_mi and other_is_mi):
```
I think this just becomes an else here

jreback replied:
or you can just remove the if as well (IOW we have covered both cases).

jreback commented on pandas/core/reshape/merge.py:
```diff
@@ -1066,6 +1066,81 @@ def _get_join_indexers(left_keys, right_keys, sort=False, how='inner',
+        right indexer
+    dropped_levels : str array
+        list of non-common levels
+    Returns
```
blank line before Returns

jreback commented on pandas/core/reshape/merge.py:
```diff
@@ -1066,6 +1066,81 @@ def _get_join_indexers(left_keys, right_keys, sort=False, how='inner',
+    """
+
+    if how == 'outer':
+        levels = join_idx.levels
```
maybe a few comments on what you are doing here (for the outer case), IOW why iterating over the dropped levels, etc.

jreback commented on pandas/tests/reshape/test_merge.py:
```diff
@@ -1344,32 +1341,194 @@ def f():
+
+        result = left.join(right, how=how)
+        tm.assert_frame_equal(result, expected)
+
```
is it possible to generically also tests the equivalency (this is just a copy-paste from above, obviously have to adapt here)

```
result = (merge(household.reset_index(), log_return.reset_index(),
                          on=['asset_id'], how='inner')
                    .set_index(['household_id', 'asset_id', 't']))
```

jreback commented on pandas/tests/reshape/test_merge.py:
```diff
@@ -1344,32 +1341,194 @@ def f():
+
+
+class TestJoinMultiMulti(object):
+    @pytest.fixture
```
I would simply put the fixture outside the class, right above TestJoinMultiMulti, then you can just pass them in as needed

jreback commented on pandas/tests/reshape/test_merge.py:
```diff
@@ -1344,32 +1341,194 @@ def f():
+                                               ('right', expected_rightj),
+                                               ('inner', expected_innerj),
+                                               ('outer', expected_outerj)])
+    def test_join_multi_multi(self, how, expected):
```
e.g. you pass left, right in here

jreback commented on pandas/tests/reshape/test_merge.py:
```diff
@@ -1344,32 +1341,194 @@ def f():
+        .set_index(['Origin', 'Destination', 'Period',
+                    'TripPurp', 'LinkType']))
+
+    @pytest.mark.parametrize('how, expected', [('left', expected_leftj),
```
these you actually DO need to evaluate (e.g. ``expected_left()``) because these are simply calling a function and are not 'parameters' (if they are in the signature they are parameters).

jreback commented on pandas/tests/reshape/test_merge.py:
```diff
@@ -1408,28 +1567,27 @@ def test_join_multi_dtypes(self, d1, d2):
 class TestMergeCategorical(object):
 
-    def test_identical(self, left):
+    @pytest.fixture
```
would leave these where they were

jreback commented on pandas/tests/reshape/test_merge.py:
```diff
@@ -1439,9 +1597,11 @@ def test_identical(self, left):
                           index=['X', 'Y_x', 'Y_y'])
         assert_series_equal(result, expected)
 
-    def test_basic(self, left, right):
```
so just leave this, it is very typical, writing ``left = self.left()`` is not.

jreback reviewed (changes requested):
can you add some tests versus empty frames (matching levels but empty). test all the hows's

harisbal commented:
@jreback I have implemented the changes but one test is failing and I'm struggling to understand why. Shall I upload so you can give it a look?

jreback commented:
@harisbal what test is failing? l

jreback commented:
can you rebase

harisbal commented:
@jreback  Any idea why the tests do not fail when I run them locally?

TomAugspurger commented:
Hmm @harisbal seems like there was a rebase snafu in https://github.com/pandas-dev/pandas/pull/16162/commits/5c27d875759f265bf79d7c9c989cf2464cab8a59. Need help fixing it up?

harisbal commented:
@TomAugspurger  thank you very much but I managed to rebase properly. I now have an issue when I try to outer join on an empty level. I cannot debug after a certain point (I assume cython comes into play) so some help would be much appreciated

jreback commented:
can you rebase. let's see if we can get this in.

harisbal commented:
Two tests fail and I can't see how to fix the issue

auvipy commented:
rebase and sqash commits or create a new pr

Commit:
Rebase

Rebase

Commit:
Rebase

Rebase

Commit:
Merge remote-tracking branch 'pandas-dev/master'

minggli committed:
Regression in make_block_same_class (tests failing for new fastparquet release) (#19434)

toobaz committed:
TST: fix test for MultiIndexPyIntEngine on 32 bit (#19440)

closes #19439

luzpaz committed:
Misc typos (#19430)

Found via `codespell -q 3`

jorisvandenbossche committed:
Change Future to DeprecationWarning for make_block_same_class (#19442)

jbrockmendel committed:
catch PerformanceWarning (#19446)

jreback committed:
CI: pin pymysql<0.8.0 (#19461)

toobaz committed:
TST: fix (other check of) test for MultiIndexPyIntEngine on 32 bit (#19455)

timhoffm committed:
remove reference to deprecated .ix from 10min.rst (#19452)

jbrockmendel committed:
remove unused (#19466)

jbrockmendel committed:
setup.py fixup, closes #19467 (#19472)

jbrockmendel committed:
Centralize Arithmetic Tests (#19471)

jbrockmendel committed:
implement bits of numpy_helper in cython where possible (#19450)

minggli committed:
[#7292] BUG: asfreq / pct_change strange behavior  (#19410)

reidy-p committed:
DEPR: Deprecate from_items (#18529)

TomAugspurger committed:
BUG: Fixed accessor for Categorical[Datetime] (#19469)

* BUG: Fixed accessor for Categorical[Datetime]

* Fixup

tommyod committed:
DOC: Spellcheck of categorical.rst and visualization.rst (#19428)

topper-123 committed:
DEPR/CLN: Remove pd.rolling_*, pd.expanding* and pd.ewm* (#18723)

* remove pd.running_*, pd.expanding_* and pd.ewm* and related code

* added test_expanding_func and test_expanding_apply

* recreate _check_ndarray inline in _check_moment_func

jbrockmendel committed:
Organize, Split, Parametrize timezones/timestamps tests (#19473)

jbrockmendel committed:
implement test_scalar_compat (#19479)

jbrockmendel committed:
Refactor out libwriters, fix references to Timestamp, Timedelta (#19413)

jreback committed:
PERF: remove use of Panel & perf in rolling corr/cov (#19257)

* PERF: remove use of Panel & perf in rolling corr/cov

closes #17917

jreback committed:
TST: fix up pandas_datareader downstream tests (#19490)

closes #18935

hexgnu committed:
BUG: fix issue with concat creating SparseFrame if not all series are sparse. (#18924)

mitchnegus committed:
updated hist documentation (#19366)

Commit:
CLN: GH19404 Changing function signature to match logic (#19425)

WBare committed:
ENH limit_area added to interpolate1d

closes #16284

hexgnu committed:
BUG: Fix problem with SparseDataFrame not persisting to csv (#19441)

* BUG: Fix problem with SparseDataFrame not persisting to csv

* FIX: Remove comment and move test with more coverage

* FIX: Flake8 issues cleanup

* Fix failing test due to blank lines

* FIX: linting errors on whitespace

* Use parametrize on test

* Move bug description to sparse header

* Add GH issue to test

* Fix linting error

TomAugspurger committed:
Added E741 to flake8 config (#19496)

jbrockmendel committed:
implement timedeltas.test_scalar_compat (#19503)

jbrockmendel committed:
Continue de-nesting core.ops (#19448)

jbrockmendel committed:
Make DateOffset.kwds a property (#19403)

jbrockmendel committed:
Fix DTI comparison with None, datetime.date (#19301)

tommyod committed:
DOC: Exposed arguments in plot.kde (#19229)

* Exposed arguments in plot.kde, added number of sample points as option

* Added a test for plot.kde with  as an integer

* Added whatsnew. Fixed flake8 errors. Used is_integer  to infer type.

* Updated scipy reference

* Added test, rewrote whatsnew, removed import

* Changed from Series to DataFrame in doc

* Fixed PEP8 errors in test file

* Fixed typo which made tests crash

TomAugspurger committed:
ENH: Array Interface and Categorical internals Refactor (#19268)

* REF: Define extension base classes

* Updated for comments

* removed take_nd
* Changed to_dense to return get_values
* Fixed docstrings, types
* Removed is_sparse

* Remove metaclasses from PeriodDtype and IntervalDtype

* Fixup form_blocks rebase

* Restore concat casting cat -> object

* Remove _slice, clarify semantics around __getitem__

* Document and use take.

* Clarify type, kind, init

* Remove base

* API: Remove unused __iter__ and get_values

* API: Implement repr and str

* Remove default value_counts for now

* Fixed merge conflicts

* Remove implementation of construct_from_string

* Example implementation of take

* Cleanup ExtensionBlock

* Pass through ndim

* Use series._values

* Removed repr, updated take doc

* Various cleanups

* Handle get_values, to_dense, is_view

* Docs

* Remove is_extension, is_bool

Remove inherited convert

* Sparse formatter

* Revert "Sparse formatter"

This reverts commit ab2f0457839fece3b3ef067f29994b42908bd037.

* Unbox SparseSeries

* Added test for sparse consolidation

* Docs

* Moved to errors

* Handle classmethods, properties

* Use our AbstractMethodError

* Lint

* Cleanup

* Move ndim validation to a method.

* Try this

* Make ExtensionBlock._holder a property

Removed ExtensionBlock.__init__

* Make _holder a property for all

* Refactored validate_ndim

* fixup! Refactored validate_ndim

* lint

reidy-p committed:
ERR: Better error msg when merging on tz-aware and tz-naive columns (#19525)

tommyod committed:
DOC: Spellcheck of enhancingperf.rst (#19516)

* Spellchecked enhancingperf, sparse

* Uppercased 'cython' to 'Cython'

* Typeset variants of numba as 'Numba', as on their page

* Updated reference to Numba docs to latest version

jschendel committed:
TST: Remove duplicate TimdeltaIndex tests (#19509)

jbrockmendel committed:
Frame specific parts of #19504 (#19512)

jbrockmendel committed:
split Timestamp tests off of 19504 (#19511)

jbrockmendel committed:
ops cleanup, named functions instead of lambdas (#19515)

reidy-p committed:
DOC: Improve replace docstring (#18100)

topper-123 committed:
DOC: minor groupby and resampler improvements (#19514)

datapythonista committed:
DEPR: Changing default of str.extract(expand=False) to str.extract(expand=True) (#19118)

jschendel committed:
TST: Remove legacy instances of _multiprocess_can_split_ (#19536)

jbrockmendel committed:
remove unused calendar options from period_helper (#19534)

discort committed:
BUG: groupby with resample using on parameter errors when selecting column to apply function

closes #17813

Author: discort <odiscort@gmail.com>

Closes #19433 from discort/fix_17813 and squashes the following commits:

2f25d40a0 [discort] Fixed bug in df.resample using 'on' parameter

toobaz committed:
TST: Fix makeIntIndex, benchmark get loc

Author: Pietro Battiston <me@pietrobattiston.it>

Closes #19483 from toobaz/test_get_loc and squashes the following commits:

51d691106 [Pietro Battiston] TST: benchmark get_loc in various cases
d424f63df [Pietro Battiston] TST: produce unsorted integer index (consistently with other types)

ppflrs committed:
DOC: Fix typo in example (#19537)

Fix typo in the example for pandas.io.formats.style.Styler.format

hexgnu committed:
BUG: don't assume series is length > 0

closes #19368

Author: Matthew Kirk <matt@matthewkirk.com>

Closes #19438 from hexgnu/segfault_memory_usage and squashes the following commits:

f9433d844 [Matthew Kirk] Use shared docstring and get rid of if condition
4ead141c0 [Matthew Kirk] Move whatsnew doc to Sparse
ae9f74d58 [Matthew Kirk] Revert base.py
cdd4141e4 [Matthew Kirk] Fix linting error
93a0c3daa [Matthew Kirk] Merge remote-tracking branch 'upstream/master' into segfault_memory_usage
207bc74d2 [Matthew Kirk] Define memory_usage on SparseArray
21ae14707 [Matthew Kirk] FIX: revert change to lib.pyx
3f52a44f6 [Matthew Kirk] Ah ha I think I got it
5e59e9cbc [Matthew Kirk] Use range over 0 <= for loops
e25158713 [Matthew Kirk] Fix failing test with indexing
27df317be [Matthew Kirk] Merge remote-tracking branch 'upstream/master' into segfault_memory_usage
7fdd03e94 [Matthew Kirk] Take out comment and use product
6bd6ddd02 [Matthew Kirk] BUG: don't assume series is length > 0

jbrockmendel committed:
TST: fix and test index division by zero

Related: #19336

Author: Brock Mendel <jbrockmendel@gmail.com>

Closes #19347 from jbrockmendel/div_zero2 and squashes the following commits:

be1e2e1b8 [Brock Mendel] move fixture to conftest
64b0c0853 [Brock Mendel] Merge branch 'master' of https://github.com/pandas-dev/pandas into div_zero2
aa969f8d2 [Brock Mendel] Merge branch 'master' of https://github.com/pandas-dev/pandas into div_zero2
000aefde0 [Brock Mendel] fix long again
9de356ab0 [Brock Mendel] revert fixture to fix test_range failures
b8cf21d3e [Brock Mendel] flake8 remove unused import
afedba98b [Brock Mendel] whatsnew clarification
b51c2e14c [Brock Mendel] fixturize
37efd5108 [Brock Mendel] make zero a fixture
965f7214e [Brock Mendel] Merge branch 'master' of https://github.com/pandas-dev/pandas into div_zero2
d648ef698 [Brock Mendel] requested edits
1ef3a6c74 [Brock Mendel] Merge branch 'master' of https://github.com/pandas-dev/pandas into div_zero2
78de1a4df [Brock Mendel] Merge branch 'master' of https://github.com/pandas-dev/pandas into div_zero2
0277d9fca [Brock Mendel] add ipython output to whatsnew
5d7e3ea0c [Brock Mendel] Merge branch 'master' of https://github.com/pandas-dev/pandas into div_zero2
ea75c3ca0 [Brock Mendel] ipython block
6fc61bd99 [Brock Mendel] elaborate docstring
ca3bf4241 [Brock Mendel] Whatsnew section
cd543497c [Brock Mendel] move dispatch_missing to core.missing
06df02a89 [Brock Mendel] py3 fix
84c74c54a [Brock Mendel] remove operator.div for py3
6acc2f78a [Brock Mendel] fix missing import
e0e89b978 [Brock Mendel] fix and and tests for divmod
969f342e1 [Brock Mendel] fix and test index division by zero

GuessWhoSamFoo committed:
DOC: Remove repeated duplicated word (#19546)

jbrockmendel committed:
centralize and split frame division tests (#19527)

jbrockmendel committed:
Fix parsing corner case closes #19382 (#19529)

jbrockmendel committed:
Collect Series timezone tests (#19541)

swyoon committed:
DOC/ERR: better error message on no common merge keys (#19427)

miker985 committed:
BUGFIX - AttributeError raised in StataReader.value_labels() (#19510)

jbrockmendel committed:
separate DatetimeIndex timezone tests (#19545)

jbandlow committed:
BUG: Fix ts precision issue with groupby and NaT (#19526)

closes #19526

Author: Jason Bandlow <jbandlow@gmail.com>

Closes #19530 from jbandlow/timestamp_float_conversion and squashes the following commits:

2fb23d673 [Jason Bandlow] merge
af37225d4 [Jason Bandlow] BUG: Fix ts precision issue with groupby and NaT (#19526)

WillAyd committed:
Cleaned up return of _get_cython_function (#19561)

reidy-p committed:
DEPR/CLN: fix from_items deprecation warnings (#19559)

jbrockmendel committed:
Implement get_day_of_year, tests (#19555)

discort committed:
fixed bug in df.aggregate passing non-existent columns (#19552)

jreback committed:
ERR: raise KeyError on invalid column name in aggregate (#19566)

xref #19552

jbrockmendel committed:
Frame ops prelims - de-duplicate, remove unused kwargs (#19522)

jreback committed:
API/BUG: .apply will correctly infer output shape when axis=1 (#18577)

closes #16353
closes #17348
closes #17437
closes #18573
closes #17970
closes #17892
closes #17602
closes #18775
closes #18901
closes #18919

cbertinato committed:
BUG: Fixes rounding error in Timestamp.floor() (#19240)

jorisvandenbossche committed:
DOC: some clean-up of the apply docs (follow-up #18577) (#19573)

jbrockmendel committed:
Remove duplicated logic from period_helper (#19540)

mroeschke committed:
CI: Run ASV on Travis for failed benchmarks (#19236)

TomAugspurger committed:
BUG: Fixed merge on dtype equal categories (#19553)

TomAugspurger committed:
PERF: Correct signature for group_nth / group_object (#19579)

xpvpc committed:
DOC: doc/source/indexing.rst says pd.df.ix is deprecated, show warning in generated doc. (#19596)

jbrockmendel committed:
Simplify argument passing in period_helper (#19550)

jbrockmendel committed:
separate numeric tests so we can isolate division by zero (#19336)

deniederhut committed:
Bug: adds support for unary plus (#19297)

TomAugspurger committed:
Ignore warnings when reading pickle files (#19580)

samghelms committed:
ENH: added an optional css id to `<table>` tags created by `frame.to_html()` (#19594)

TomAugspurger committed:
CI: Fixed NumPy pinning in conda-build (#19575)

* CI: Fixed NumPy pinning in conda-build

* Unpin NumPy

Quite install

* Pin numpy

* Unpin everywhere else

* Build vs. 1.11

* remove one more pin

* Remove one more pin

* bump pyarrow

TomAugspurger committed:
API: Default ExtensionArray.astype (#19604)

* API: Default ExtensionArray.astype

(cherry picked from commit 943a915562b72bed147c857de927afa0daf31c1a)

* Py2 compat

* Moved

* Moved dtypes

WillAyd committed:
PERF: Cythonize Groupby Rank (#19481)

WillAyd committed:
Consolidate nth / last object Groupby Implementations (#19610)

jreback committed:
Revert "Consolidate nth / last object Groupby Implementations (#19610)"

This reverts commit d4730e65fd2dd6235158930f756e1f1afd298488.

datajanko committed:
ENH: df.assign accepting dependent **kwargs (#14207) (#18852)

elrubio committed:
Fix left join turning into outer join (#19624)

jbrockmendel committed:
function for frequently repeated tzconversion code (#19625)

jschendel committed:
API: Allow ordered=None in CategoricalDtype (#18889)

jbrockmendel committed:
order of exceptions in array_to_datetime (#19621)

WillAyd committed:
Consolidated Groupby nth / last object Templates (#19635)

jbrockmendel committed:
Continue porting period_helper to cython (#19608)

jbrockmendel committed:
fix overflows in Timestamp.tz_localize near boundaries (#19626)

jbrockmendel committed:
move shift_months test to test_arithmetic (#19636)

jbrockmendel committed:
move libfreqs and liboffsets tests to test_tslibs, move parsing tests, with to_datetime test moved to test_tools (#19638)

jbrockmendel committed:
Fix uncaught OutOfBounds in array_to_datetime (#19612)

jbrockmendel committed:
test_astype portion of #19627 (#19637)

jbrockmendel committed:
move timedelta test_astype test (#19639)

jbrockmendel committed:
Organize PeriodIndex tests (#19641)

Licht-T committed:
TST: Add to_csv test when writing the single column CSV (#19091)

Closes gh-18676

jorisvandenbossche committed:
TST: set multi_statement flag for pymysql tests (#19619)

* Revert "CI: pin pymysql<0.8.0 (#19461)"

This reverts commit 44bbd5a4d33643c9270bbefd7419f45aecaa4667.

* Enable multi-statements for pymysql connection

jbrockmendel committed:
move array_to_datetime timests (#19640)

hexgnu committed:
BUG: assign doesnt cast SparseDataFrame to DataFrame (#19178)

jreback committed:
TST: placement of network error catching in s3 tests (#19645)

jbrockmendel committed:
De-duplicate masking/fallback logic in ops (#19613)

TomAugspurger committed:
REF: Internal / External values (#19558)

* REF/Clean: Internal / External values

* Move to index base

* Cleanup unique handling

* Simplify object concat

* Use values for intersection

I think eventually we'll want to ndarray_values for this, but it'll
require a bit more work to support. Currently, using ndarary_values
causes occasional failures on categorical.

* hmm

* Additional testing

* More tests

* ndarray_values

* API: Default ExtensionArray.astype

(cherry picked from commit 943a915562b72bed147c857de927afa0daf31c1a)
(cherry picked from commit fbf0a0672380e210d3cb3c527fa8045a204d81be)

* Simplify concat_as_object

* Py2 compat

(cherry picked from commit b20e12cae68dd86ff51597464045656763d369f7)

* Set-ops ugliness

* better docstrings

* tolist

* linting

* Moved dtypes

(cherry picked from commit d1362271bca8a7b183f3241e5c2f040c422118b8)

* clean

* cleanup

* NumPy compat

* Use base _values for CategoricalIndex

* Update dev docs

* cleanup

* Linting

* Precision in tests

* Push _ndarray_values to ExtensionArray

Now IndexOpsMixin._ndarray_values will dispatch all the way down to the EA.
Subclasses like Categorical can override it as they see fit.

* Clean up tolist

* Move test locations

* Fixed test

* REF: Update per comments

* lint

* REF: Use _values for size and shape

* PERF: Implement size, shape for IntervalIndex

* PERF: Avoid materializing values for PeriodIndex shape, size

* Cleanup

* Override nbytes

jorisvandenbossche committed:
DOC: ignore Panel deprecation warnings during doc build (#19663)

Carreau committed:
DOC: fix IPython spelling (#19683)

It's upper case I and P (or full lower case),

WillAyd committed:
Explicitly set dtype of np.lexsort in group_rank (#19679)

mroeschke committed:
BUG: Do not round DatetimeIndex nanosecond precision when iterating (#19628)

AaronCritchley committed:
COMPAT-18589: Supporting axis in Series.rename (#18923)

hexgnu committed:
Performance increase rolling min max (#19549)

jbrockmendel committed:
tests for tslibs.conversion and tslibs.timezones (#19642)

tommyod committed:
Spellchecked io.rst (#19660)

TomAugspurger committed:
CI: Move conda build and ASV check to cron job (#19698)

* CI: Remove ASV run

* CI: Removed conda build test

* Removed asv.sh step

* Removed the ASV matrix item

WillAyd committed:
GroupBy Rank SegFault Fix - astype instead of view (#19701)

* Use astype instead of view for lexsort upcasting

* Added copy=False to astype in group_rank

giba0 committed:
DOC: Ambiguous description in to_parquet engine documentation  (#19669)

No-Stream committed:
ENH: groupby().is_monotonic_increasing #17015 (#17453)

cbrnr committed:
DOC: improve docs to clarify MultiIndex indexing (#19507)

jbrockmendel committed:
add missing args, make kwarg explicit (#19691)

jbrockmendel committed:
remove usages of _get_na_value (#19692)

jbrockmendel committed:
TST: Parametrize PeriodIndex tests (#19659)

* fixup formatting

* parametrize PeriodIndex tests

* fixup typo

* put lists of params at module level

* make fixtures

* docstrings for fixtures

* requested docstring

tommyod committed:
DOC: Updated tutorials with additional info, new version and added some video tutorials (#19748)

jbrockmendel committed:
collect index formatting tests (#19661)

jbrockmendel committed:
finish off tests.tseries.test_timezones (#19739)

jbrockmendel committed:
dispatch frame methods to series versions instead of re-implementing masking etc (#19611)

WillAyd committed:
Removed if...else for K > 1 (#19734)

jbrockmendel committed:
Dispatch categorical Series ops to Categorical (#19582)

datapythonista committed:
DOC/BLD: Pinning sphinx to 1.5, as 1.7 has been released and it's incompatible with vendored numpydoc (#19743)

ZhuBaohe committed:
DOC: correct merge_asof example (#19737)

tacaswell committed:
FIX: const-correctness in numpy helpers (#19749)

In python 3.7 the return type of PyUnicode_AsUTF8 changed
from (char *) to (const char *).

PyUnicode_FromString also takes (const char *) as input, also be
explicit about that.

https://bugs.python.org/issue28769
commit 2a404b63d48d73bbaa007d89efb7a01048475acd in cpython

jorisvandenbossche committed:
DOC/BLD: update vendored IPython.sphinxext version (#19765)

updated to commit cc353b25b0fff58e4ed13899df9b3c8153df01d9 from ipython/ipython

topper-123 committed:
add test for numpy ops, esp. nanmin/max bug for np<1.13 (#19753)

ZhuBaohe committed:
DOC: correct Period.strftime exsample (#19758)

yarikoptic committed:
ENH: fake http proxy in case of --skip-network testing (#19757)

ZhuBaohe committed:
DOC: correct Panel.apply exsample (#19766)

jbrockmendel committed:
split off scalar tests to submodules (#19752)

jreback committed:
CI: remove PIP & old conda build in favor of pandas-ci buildsx (#19775)

jbrockmendel committed:
ENH: implement Timedelta.__mod__ and __divmod__ (#19755)

jbrockmendel committed:
Fix Timedelta floordiv, rfloordiv with offset, fix td64 return types (#19770)

jbrockmendel committed:
Reduce redirection in ops (#19649)

hexgnu committed:
Fix the non cython build for cpp extensions (#19707)

jreback committed:
DOC: whatsnew typo cleanup

jorisvandenbossche committed:
DOC: fix various warnings and errors in the docs (from deprecations/api changes) (#19763)

jbrockmendel committed:
Split+Parametrize Timedelta tests (#19736)

mikekutzma committed:
BUG: GH19458 fixes precision issue in TimeDelta.total_seconds() (#19783)

jreback committed:
DOC: whatsnew cleanups

jreback committed:
DOC: typos in whatsnew

obilodeau committed:
DOC: added a reference to DataFrame assign in concatenate section of merging (#18665)

jbrockmendel committed:
Sparse Ops Cleanup (#19782)

jbrockmendel committed:
BUG: fix Period.asfreq conversion near datetime(1, 1, 1) (#19650)

ZhuBaohe committed:
DOC: correct Series.searchsorted example (#19784)

antquinonez committed:
DOC: Edit installation instructions for clarity. (#19798)

yarikoptic committed:
BF: Skip test_read_excel_parse_dates if no xlwt which is used in to_excel (#19803)

EricChea committed:
DEPR: Add deprecation warning for factorize() order keyword (#19751)

NoahTheDuke committed:
BUG: drop_duplicates not raising KeyError on missing key (#19730)

jreback committed:
ASV: excel asv occasional failure (#19811)

closes #19779

EricChea committed:
DOC: Add example of how to preserve order of columns with usecols. (#19746)

* Add example of how to preserve order of columns with usecols.

* Encase usecols in double back ticks for consistency. Change column names from numeric to string.

* Add line to separate examples.

jbrockmendel committed:
TST: move more series tests to test_arithmetic (#19794)

jbrockmendel committed:
Fix name setting in DTI/TDI __add__ and __sub__ (#19744)

jbrockmendel committed:
parametrize a whole mess of tests (#19785)

jreback committed:
DEPR: remove pandas.core.common is_* (#19769)

HagaiHargil committed:
DOC: Clarify and add fill_value example in arithmetic ops (#19675)

mehemken committed:
DOC: added plotting module to the api reference docs (#19780)

TomAugspurger committed:
API: Validate keyword arguments to fillna (#19684)

jbrockmendel committed:
Fix Index __mul__-like ops with timedelta scalars (#19333)

datapythonista committed:
DOC: Improving code quality of doc/make.py, PEP-8, refactoring and removing unused commands (#19631) (#19634)

max-sixty committed:
DOC: RangeIndex as default index (#19781)

kylebarron committed:
Update df.to_stata() docstring (#19818)

ZhuBaohe committed:
DOC: correct Series.reset_index example (#19832)

jbrockmendel committed:
implement add_offset_array for PeriodIndex (#19826)

reidy-p committed:
ENH: Add columns parameter to from_dict (#19802)

jbrockmendel committed:
fix Timedelta.__mul__(NaT) (#19819)

jbrockmendel committed:
Fix rfloordiv return type, un-xfail Timedelta tests (#19820)

jschendel committed:
BUG: Fix qcut with NaT present (#19833)

jorisvandenbossche committed:
CI: Align pep8speaks config with setup.cfg (#19841)

datapythonista committed:
DOC: Making doc/source/conf.py pass PEP-8, and added to lint (#19839)

topper-123 committed:
Let initialisation from dicts use insertion order for py>=36, part I (#19830)

spacesphere committed:
BUG: Fix MultiIndex .loc with all numpy arrays (#19772)

ZhuBaohe committed:
DOC: correct min_count param docstring (#19836)

jbrockmendel committed:
Continue porting period_helper; fix leftover asfreq bug (#19834)

jbrockmendel committed:
BUG: fix index op names and pinning (#19723)

tommyod committed:
DOC: Spellcheck of gotchas.rst (FAQ page) (#19747)

TomAugspurger committed:
ENH: Allow storing ExtensionArrays in containers (#19520)

* ENH: non-interval changes

* COMPAT: py2 Super

* BUG: Use original object for extension array

* Consistent boxing / unboxing

NumPy compat

* 32-bit compat

* Add a test array

* linting

* Default __iter__

* Tests for value_counts

* Implement value_counts

* Py2 compat

* Fixed dropna

* Test fixups

* Started setitem

* REF/Clean: Internal / External values

* Move to index base

* Setitem tests, decimal example

* Compat

* Fixed extension block tests.

The only "API change" was that you can't just inherit from
NonConsolidatableMixin, which is OK since

1. it's a mixin
2. geopandas also inherits from Block

* Clarify binop tests

Make it clearer which bit might raise

* TST: Removed ops tests

* Cleanup unique handling

* Simplify object concat

* Use values for intersection

I think eventually we'll want to ndarray_values for this, but it'll
require a bit more work to support. Currently, using ndarary_values
causes occasional failures on categorical.

* hmm

* More failing tests

* remove bad test

* better setitem

* Dropna works.

* Restore xfail test

* Test Categorical

* Xfail setitem tests

* TST: Skip JSON tests on py2

* Additional testing

* More tests

* ndarray_values

* API: Default ExtensionArray.astype

(cherry picked from commit 943a915562b72bed147c857de927afa0daf31c1a)
(cherry picked from commit fbf0a0672380e210d3cb3c527fa8045a204d81be)

* Simplify concat_as_object

* Py2 compat

(cherry picked from commit b20e12cae68dd86ff51597464045656763d369f7)

* Set-ops ugliness

* better docstrings

* tolist

* linting

* Moved dtypes

(cherry picked from commit d1362271bca8a7b183f3241e5c2f040c422118b8)

* clean

* cleanup

* NumPy compat

* Use base _values for CategoricalIndex

* Update dev docs

* cleanup

* cleanup

(cherry picked from commit 242562108b099b4e7a205541ee15b9272dcb5265)

* cleanup

* Linting

* Precision in tests

* Linting

* Move to extension

* Push _ndarray_values to ExtensionArray

Now IndexOpsMixin._ndarray_values will dispatch all the way down to the EA.
Subclasses like Categorical can override it as they see fit.

* Clean up tolist

* Move test locations

* Fixed test

* REF: Update per comments

* lint

* REF: Use _values for size and shape

* PERF: Implement size, shape for IntervalIndex

* PERF: Avoid materializing values for PeriodIndex shape, size

* Cleanup

* Override nbytes

* Remove unused change

* Docs

* Test cleanpu

* Always set PANDAS_TESTING_MODE

* Revert "Always set PANDAS_TESTING_MODE"

This reverts commit a312ba5c59c2e96854a286bde74d7fd4562afbf8.

* Explicitly catch warnings or not

* fastparquet warnings

* Unicode literals strikes again.

Only catch fp warning for newer numpy

* Restore circle env var

* More parquet test catching

* No stacklevel

* Lower bound on FP

* Exact bound for FP

* Don't use fastpath for ExtensionBlock make_block

* Consistently use _values

* TST: Additional constructor tests

* CLN: de-nested a bit

* _fill_value handling

* Handle user provided dtype in constructors.

When the dtype matches, we allow it to proceed.

When the dtype would require coercion, we raise.

* Document ExtensionBlock._maybe_coerce_values

Also changes to use _values as we should

* Created ABCExtensionArray

* TST: Tests for is_object_dtype and is_string_dtype and EAs

* fixup! Handle user provided dtype in constructors.

* Doc for setitem

* Split base tests

* Revert test_parquet changes

* API: Removed _fill_value from the interface

* Push coercion to extension dtype till later

* Linting

* ERR: Better error message for coercion to 3rd party dtypes

* CLN: Make take_nd EA aware

* Revert sparse changes

* Other _typ for ABCExtensionArray

* Test cleanup and expansion.

Tests for concating and aligning frames

* Copy if copy

* TST: remove self param for fixture

* Remove unnescessary EA handling in Series ctor

* API: Removed value_counts

Moved setitem notes to comment

* More doc notes

* Handle expanding a DataFrame with an EA

* Added ExtensionDtype.__eq__

Support for astype

* linting

* REF: is_dtype_equal refactor

Moved from PandasExtensionDtype to ExtensionDtype with one modification:
catch TypeError explicitly.

* Remove reference to dtype being a class

* move

* Moved sparse check to take_nd

* Docstring

* Split tests

* Revert index change

* Copy changes

* Simplify EA implementation names

comments for object vs. str missing values

* Linting

jbrockmendel committed:
Separate TimedeltaIndex mul/div tests (#19848)

luzpaz committed:
DOC: misc. typos (#19876)

Found via `codespell -q 3 -I ../pandas-whitelist.txt`
Where whitelists consists of:
```
ans
behaviour
doubleclick
indicies
initialise
initialised
initialising
nd
resetted
splitted
thru
valu
```

jorisvandenbossche committed:
DOC: remove deprecated from_items from dsintro docs (#19837)

jbrockmendel committed:
De-duplicate add_offset_array methods (#19835)

topper-123 committed:
Let initialisation from dicts use insertion order for python >= 3.6 (part II) (#19859)

cbertinato committed:
BUG: fix Series constructor for scalar and Categorical dtype (#19717)

jayfoad committed:
Raise OptionError instead of KeyError in __getattr__. Fixes  #19789. (#19790)

jaumebonet committed:
Keep subclassing in apply (#19823)

TomAugspurger committed:
REF: Base class for all extension tests (#19863)

tommyod committed:
DOC: Updated links to 2 tutorials in tutorials.rst (#19857)

jbrockmendel committed:
templatize timedelta arith ops (#19871)

jreback commented:
@harisbal sorry let this drop. can you rebase and we can get this in.

jreback committed:
COMPAT: fixup decimal extension for indexing compat (#19882)

jreback committed:
CI: pin jemalloc=4.5.0.poast for 2.7 build per (#19888)

https://issues.apache.org/jira/browse/ARROW-2208

WillAyd committed:
Cythonized GroupBy Fill (#19673)

WillAyd committed:
Fixed pct_change with 'fill_method' returning NaN instead of 0 (#19875)

jbrockmendel committed:
Use pandas_datetimestruct instead of date_info (#19874)

jbrockmendel committed:
Fix+test DTI/TDI/PI add/sub with ndarray[datetime64/timedelta64] (#19847)

jjames34 committed:
Fixed issue with leftover test.json file (#19879)

minggli committed:
ENH: ISO8601-compliant datetime string conversion in `iterrows()` and Series construction. (#19762)

minggli committed:
parameterize test_pct_change_periods_freq (#19897)

Commit:
DOC: Make API reference intro section concise (#19846)

jorisvandenbossche committed:
DOC/BLD: unpin sphinx to use sphinx 1.7 (#19687)

jorisvandenbossche committed:
DOC: fix numpydoc section titles in misc plotting docstrings (#19899)

ColCarroll committed:
DOC: small typo fix (#19901)

jbrockmendel committed:
cleanup order of operations kludges (#19895)

jbrockmendel committed:
make ops.add_foo take just class (#19828)

WillAyd committed:
Test Decorators and Better Pytest Integration in 'test_excel' (#19829)

cbertinato committed:
BUG: Fix Series constructor for Categorical with index (#19714)

jaumebonet committed:
CLN: Remove Series._from_array (#19893)

Commit:
DOC fix incorrect example in DataFrame.to_dict docstring. Close GH19868 (#19915)

jbrockmendel committed:
handle NaT add/sub in one place (#19903)

WillAyd committed:
ASV: Added seek to buffer to fix xlwt asv failure (#19926)

* Added seek to buffer to fix xlwt asv failure

* Added conditional to check for seek on xlrd object

TomAugspurger committed:
TST: Debug flaky plotting test (#19925)

Commit:
Merge remote-tracking branch 'pandas-dev/master'

harisbal committed:
Rebase

Commit:
Merge branch 'master' of https://github.com/pandas-dev/pandas

Commit:
Merge branch 'master' into multi-index-merge

# Conflicts:
#	doc/source/merging.rst
#	doc/source/whatsnew/v0.23.0.txt
#	pandas/core/frame.py
#	pandas/core/generic.py
#	pandas/core/indexes/base.py
#	pandas/core/ops.py
#	pandas/core/reshape/merge.py
#	pandas/plotting/_misc.py
#	pandas/tests/reshape/merge/test_merge.py

harisbal commented:
This pull request is dirty beyond repair. I'll squash and start a new one
